File: leetcode/medium/3133_minEnd.py
from typing import List
from json import dumps
import heapq


'''

4 = 100
5 iteration
00100
00101
00110
00111
01100

changes in zeros:
001
010
011
100

final result of combinations equal to n-1
if we know this trick, we can put bits from n-1 into zeros place

'''

class Solution:
    def minEnd(self, n: int, x: int) -> int:
        print('x', bin(x))


        combArr = list(bin(n-1)[2:])
        xArr = list(bin(x)[2:])
        rem = ['0']*(64 - len(xArr))
        xArr = rem + xArr

        print(
            'combArr', combArr,
            'xArr', xArr,

        )

        id = 63
        for i in range(len(combArr) - 1, -1, -1):
            while xArr[id] != '0':
                id -= 1
            xArr[id] = combArr[i]
            id -= 1

        res = int('0b'+''.join(xArr), 2)
        print('xArr', '0b'+''.join(xArr))
        print('xArr', int('0b'+''.join(xArr), 2))

        return res


def test ():
    params = [
        {
            'input': [3, 4],
            'output': 6,
        },
        {
            'input': [2, 7],
            'output': 15,
        },
        {
            'input': [5, 4],
            'output': 12,
        },
        {
            'input': [6715154, 7193485],
            'output': 55012476815,
        },  
    ]
    solution = Solution()

    for param in params:
        n, x = param['input']
        result = solution.minEnd(n, x)
        print(
            'SUCCESS' if dumps(result) == dumps(param['output']) else 'ERROR',
            'input', param['input'],
            'output', param['output'],
            'result', result,
            '\n',
        )


if __name__ == '__main__':
    test()

File: leetcode/medium/3133_minEnd.js
/**
 * @param {number} n
 * @param {number} x
 * @return {number}
 */
var minEnd_1 = function(n, x) {
    console.log(
        'x: \n',
        x, parseInt(x).toString(2),

    );

    const arr = [x];
    const bitArr = [];
    let rem = x;
    while (rem !== 0) {
        bitArr.push(rem % 2);
        rem = Math.floor(rem / 2);
    }

    console.log(
        'bitArr', bitArr,
    );

    for (let i = 1; i < n; ++i) {
        let zero = -1;
        for (let j = 0; j < bitArr.length; ++j) {
            if (bitArr[j] === 0) {
                zero = j;
                break;
            }
        }

        if (zero != -1) {
            bitArr[zero] = 1;
            arr.push(parseInt([...bitArr].reverse().join(''), 2));
        }

        console.log(
            'bitArr', bitArr,
        );
    }

    console.log(
        'arr', arr,
    );


    return 0;
};

var minEnd_2 = function(n, x) {
    let last = x;

    for (let i = 1; i < n; ++i) {
        let b = last + 1;
        while ((b & x) !== x) {
            b += 1;
        }
        last = b;
    }

    return last;
};

var minEnd = function(n, x) {
    let last = x;

    for (let i = 1; i < n; ++i) {
        last = (last + 1) | x;
    }

    return last;
};




const test = () => {
    const params = [
        // [4, 5, 6] = [100, 101, 110] = 100 & 101 & 110 = 100
        {
            input: [3, 4],
            output: 6,
        },
        // [7, 15] = [111, 1111] = 0111 & 1111 = 111
        {
            input: [2, 7],
            output: 15,
        },
        {
            input: [4, 1],
            output: 7,
        },
        {
            input: [6715154, 7193485],
            output: 55012476815,
        },
    ];

    for (const { input, output } of params) {
        const result = minEnd(...input);
        const message = `
            INPUT: ${input}
            OUTPUT: ${output}
            RESULT: ${result}
            `;

        if (result === output) {
            console.log(
                'SUCCESS: \n', message,
            );
        } else {
            console.error('ERROR: \n', message);
        }
    }
};

test();
